packages/runtime/container-runtime/src/opLifecycle/outbox.ts

```diff
@@ -191,22 +219,85 @@ export class Outbox {
 			this.attachFlowBatch.contentSizeInBytes >=
 			this.params.config.compressionOptions.minimumBatchSizeInBytes
 		) {
-			this.flushInternal(this.attachFlowBatch.popBatch());
+			this.flushInternal(this.attachFlowBatch);
 		}
 	}
 
 	public flush() {
-		this.flushInternal(this.attachFlowBatch.popBatch());
-		this.flushInternal(this.mainBatch.popBatch());
+		if (this.isContextReentrant()) {
```

Comment:
It feels like this condition should not be possible, given that we do not expose flush() on API surface.
If that's true, I wonder if assert is enough here? Trying to hit a balance between enforcing invariants and code being not very verbose (but not sure if assert for sure closes container here).

Reply:
There is an end-to-end test showcasing how this can be possible.

Reply:
https://github.com/microsoft/FluidFramework/pull/16008/files#diff-562098eab98c8630e03564af13e114fb1f62acc6ca629b3fb1a255618131db09R275